### PR TITLE
ML-4171: Redis target: move to one redis-key per entity record

### PR DESCRIPTION
**Pull request by alxtkr77**

### CODE REVIEW

File: storey/redis_driver.py [hunk 1]
```diff
@@ -526,25 +547,35 @@ async def _load_aggregates_by_key(self, container, table_path, key):
...
+            values = {}
+            while True:
+                cursor, v = await RedisDriver.asyncify(self.redis.hscan)(
+                    redis_key, cursor, match=f"{self.AGGREGATION_ATTRIBUTE_PREFIX}*"
+                )
+                values.update(v)
+                if cursor == 0:
+                    break
+        except redis.ResponseError as e:
+            raise RedisError(f"Failed to get key {redis_key}. Response error was: {e}")
```
Comment (gtopper) [inline storey/redis_driver.py:536]: You can add `from e`:
```suggestion
            raise RedisError(f"Failed to get key {redis_key}. Response error was: {e}") from e
```
In theory that means you can remove the "Response error was" part, but in practice, chained exceptions are often not printed where you might expect them to be, so up to you whether to drop that or not.

File: storey/redis_driver.py [hunk 2]
```diff
@@ -526,25 +547,35 @@ async def _load_aggregates_by_key(self, container, table_path, key):
...
-        for aggr_key in all_aggr_keys:
-            aggr_key = RedisDriver.convert_to_str(aggr_key)
-            value = await RedisDriver.asyncify(self.redis.hget)(aggr_key_prefix, aggr_key)
+        redis_key = self._static_data_key(redis_key_prefix)
+        try:
+            cursor = 0
+            values = {}
+            while True:
+                cursor, v = await RedisDriver.asyncify(self.redis.hscan)(
+                    redis_key, cursor, match=f"{self.AGGREGATION_ATTRIBUTE_PREFIX}*"
```
Comment (gtopper) [inline storey/redis_driver.py]: Maybe we can parallelize these calls? i.e. with `asyncio.create_task` and `asyncio.gather`.

File: storey/redis_driver.py [hunk 3]
```diff
@@ -526,25 +547,35 @@ async def _load_aggregates_by_key(self, container, table_path, key):
...
-        aggr_key_prefix = f"{redis_key_prefix}:{RedisDriver.AGGREGATION_ATTRIBUTE_PREFIX}"
-        all_aggr_keys = self.redis.hkeys(aggr_key_prefix)
-        for aggr_key in all_aggr_keys:
-            aggr_key = RedisDriver.convert_to_str(aggr_key)
-            value = await RedisDriver.asyncify(self.redis.hget)(aggr_key_prefix, aggr_key)
+        redis_key = self._static_data_key(redis_key_prefix)
+        try:
+            cursor = 0
+            values = {}
+            while True:
```
Comment (gtopper) [inline storey/redis_driver.py:530]: This pattern appears 3 times by my count (not including tests). I think it can be extracted out.

**Review by gtopper — COMMENTED** [review]
Reviewed incidental aspect of this PR – not the redis changes themselves.

File: storey/redis_driver.py (hunk 2 above)
Comment (alxtkr77) [reply storey/redis_driver.py]: We cannot, since 'cursor' is returned and used in following call

File: storey/redis_driver.py (hunk 1 above)
Comment (alxtkr77) [reply storey/redis_driver.py:536]: ok

File: storey/redis_driver.py (hunk 3 above)
Comment (alxtkr77) [reply storey/redis_driver.py:530]: will refactor

File: integration/test_flow_integration.py [hunk 4]
```diff
@@ -111,7 +111,16 @@ def _get_redis_kv_all_attrs(setup_teardown_test: ContextForTests, key: str):
     hash_key = RedisDriver.make_key("storey-test:", table_name, key)
     redis_key = RedisDriver._static_data_key(hash_key)
     redis_fake_server = setup_teardown_test.redis_fake_server
-    values = get_redis_client(redis_fake_server=redis_fake_server).hgetall(redis_key)
+
+    cursor = 0
+    values = {}
+    while True:
+        cursor, v = get_redis_client(redis_fake_server=redis_fake_server).hscan(
+            redis_key, cursor, match=f"[^{chr(0x1)}]*"
```
Comment (assaf758) [inline integration/test_flow_integration.py]: use INTERNAL_FIELD_PREFIX

Comment (assaf758) [inline integration/test_flow_integration.py]: use  INTERNAL_FIELD_PREFIX

**Review by assaf758 — CHANGES_REQUESTED** [review]

**Review by assaf758 — APPROVED** [review]

**Pull request merged**